File: WhatsApp-agents/baileys/src/contracts/env.test.ts
//Mariano Montini ('bosque', 'bosquestudio')
import { describe, expect, it } from 'vitest'
import { parseEnv } from './env.js'

describe('parseEnv', () => {
  // Accepts empty env - defaults auth dir, port, menu text, delay range.
  it('accepts defaults and pins delay and text menu', () => {
    const result = parseEnv({})
    expect(result.ok).toBe(true)
    if (result.ok) {
      expect(result.env.BAILEYS_AUTH_DIR).toBe('./auth_info_baileys')
      expect(result.env.PORT).toBe(3003)
      expect(result.env.HUMANIZE_MIN_MS).toBe(20_000)
      expect(result.env.HUMANIZE_MAX_MS).toBe(45_000)
      expect(result.env.BAILEYS_MENU_MODE).toBe('text')
    }
  })

  // Rejects inverted delay range - max must be >= min.
  it('fails when HUMANIZE_MAX_MS is less than HUMANIZE_MIN_MS', () => {
    const result = parseEnv({
      HUMANIZE_MIN_MS: '100',
      HUMANIZE_MAX_MS: '50',
    })
    expect(result.ok).toBe(false)
    if (!result.ok) {
      expect(result.error).toContain('HUMANIZE_MAX_MS')
    }
  })

  // Production delay floor - NODE_ENV=production rejects 0 ms humanize bypass.
  it('rejects HUMANIZE_MIN_MS 0 when NODE_ENV is production', () => {
    const result = parseEnv({
      NODE_ENV: 'production',
      HUMANIZE_MIN_MS: '0',
      HUMANIZE_MAX_MS: '0',
    })
    expect(result.ok).toBe(false)
    if (!result.ok) {
      expect(result.error).toMatch(/HUMANIZE_MIN_MS|HUMANIZE_MAX_MS/)
    }
  })

  // Non-production delay bypass - tests may use 0/0 with injectable sleep.
  it('accepts HUMANIZE delays of 0 when NODE_ENV is test', () => {
    const result = parseEnv({
      NODE_ENV: 'test',
      HUMANIZE_MIN_MS: '0',
      HUMANIZE_MAX_MS: '0',
    })
    expect(result.ok).toBe(true)
    if (result.ok) {
      expect(result.env.HUMANIZE_MIN_MS).toBe(0)
      expect(result.env.HUMANIZE_MAX_MS).toBe(0)
    }
  })

  // Rejects invalid coupon URL - fail fast on outbound media contract.
  it('fails fast when COUPON_MEDIA_URL is not a URL', () => {
    const result = parseEnv({ COUPON_MEDIA_URL: 'not-a-url' })
    expect(result.ok).toBe(false)
    if (!result.ok) {
      expect(result.error).toContain('COUPON_MEDIA_URL')
    }
  })
})

File: WhatsApp-agents/whatsmeow/internal/contracts/env_test.go
package contracts

import "testing"

func TestParseEnvDefaults(t *testing.T) {
	result := ParseEnv(map[string]string{})
	if !result.OK {
		t.Fatalf("expected ok, got %s", result.Error)
	}
	if result.Env.StoreDir != "./whatsmeow_store" {
		t.Fatalf("store dir: %s", result.Env.StoreDir)
	}
	if result.Env.Port != 3004 {
		t.Fatalf("port: %d", result.Env.Port)
	}
	if result.Env.HumanizeMinMS != 20_000 || result.Env.HumanizeMaxMS != 45_000 {
		t.Fatalf("delays: %d %d", result.Env.HumanizeMinMS, result.Env.HumanizeMaxMS)
	}
	if result.Env.MenuMode != "text" {
		t.Fatalf("menu: %s", result.Env.MenuMode)
	}
}

func TestParseEnvRejectsInvertedRange(t *testing.T) {
	result := ParseEnv(map[string]string{"HUMANIZE_MIN_MS": "100", "HUMANIZE_MAX_MS": "50"})
	if result.OK {
		t.Fatal("expected fail")
	}
	if result.Error == "" || !contains(result.Error, "HUMANIZE_MAX_MS") {
		t.Fatalf("error: %s", result.Error)
	}
}

func TestParseEnvProductionRejectsEmptyWebhookSecret(t *testing.T) {
	result := ParseEnv(map[string]string{
		"GO_ENV": "production",
	})
	if result.OK {
		t.Fatal("expected production secret fail")
	}
	if !contains(result.Error, "WHATSMEOW_WEBHOOK_SECRET") {
		t.Fatalf("error: %s", result.Error)
	}
}

func TestParseEnvProductionAcceptsSecret(t *testing.T) {
	result := ParseEnv(map[string]string{
		"GO_ENV":                   "production",
		"WHATSMEOW_WEBHOOK_SECRET": "prod-secret",
	})
	if !result.OK {
		t.Fatalf("expected ok: %s", result.Error)
	}
}

func TestParseEnvProductionRejectsZeroDelay(t *testing.T) {
	result := ParseEnv(map[string]string{
		"NODE_ENV": "production", "HUMANIZE_MIN_MS": "0", "HUMANIZE_MAX_MS": "0",
	})
	if result.OK {
		t.Fatal("expected production floor fail")
	}
}

func TestParseEnvTestAllowsZeroDelay(t *testing.T) {
	result := ParseEnv(map[string]string{
		"GO_ENV": "test", "HUMANIZE_MIN_MS": "0", "HUMANIZE_MAX_MS": "0",
	})
	if !result.OK {
		t.Fatalf("expected ok: %s", result.Error)
	}
	if result.Env.HumanizeMinMS != 0 || result.Env.HumanizeMaxMS != 0 {
		t.Fatal("expected 0/0")
	}
}

func TestParseEnvInvalidCouponURL(t *testing.T) {
	result := ParseEnv(map[string]string{"COUPON_MEDIA_URL": "not-a-url"})
	if result.OK {
		t.Fatal("expected fail")
	}
	if !contains(result.Error, "COUPON_MEDIA_URL") {
		t.Fatalf("error: %s", result.Error)
	}
}

func contains(s, sub string) bool {
	return len(s) >= len(sub) && (s == sub || len(sub) == 0 || search(s, sub))
}

func search(s, sub string) bool {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}
